### ISSUE TITLE
[@typescript-eslint/no-unsafe-return] A value is not assignable to itself

### ISSUE THREAD

**Issue by Richienb** [typescript-eslint/typescript-eslint#3632]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

https://github.com/Richienb/typescript-eslint-no-unsafe-return-reproduction

**Expected Result**

No lint errors

**Actual Result**

>  11:4   error    Unsafe return of type `any` from function with return type `ReturnType<ProvidedFunction>`  @typescript-eslint/no-unsafe-return
>  16:3   error    Unsafe return of an `any` typed value                                                      @typescript-eslint/no-unsafe-return

**Additional Info**

See https://github.com/sindresorhus/mem/pull/76#discussion_r667979630 for context

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.3` |
| `@typescript-eslint/parser`        | `4.28.3` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.30.0` |
| `node`                             | `14.16.1` |

You can also [see the lockfile](https://github.com/Richienb/typescript-eslint-no-unsafe-return-reproduction/blob/53ae435dfdd4ecaf98cd4bc8fb3c4c90ccfbc619/yarn.lock).

**Comment by bradzacher** [comment]
The return type of `function_.apply(this, arguments_)` is `any` (thus the type of `result` is `any`).
The return type of the function is explicitly marked as `ReturnType<ProvidedFunction>`.

Thus you are returning an explicit any into a generic type - which is an error.
The rule is working as expected.

The point to understand here is that `ReturnType<ProvidedFunction>` ***is not the same as*** `any`.
Why? Because `ProvidedFunction` is a generic type. Even though it has a generic constraint which specifies it as "at least a function that returns `any`", in reality it doesn't have to (and likely won't) explicitly return `any`. Thus `ReturnType<ProvidedFunction>` is as much resolution as TS can actually do - it's an opaque, unresolvable type that is distinct from every other type.

Thus it is not `any`, thus you are returning `any` into a non-`any` position.
So the rule is working as expected.

----

So why does TS resolve the the `.apply` return value to to `any` instead of something like `ReturnType<ProvidedFunction>`?
Well because of how TS assigns the types for the `.apply` call.

[This is the signature TS is using for your `.apply` call on L14](
https://github.com/microsoft/TypeScript/blob/c17fd8c479c58795253b25e7272ae9072d68a9a5/lib/lib.es5.d.ts#L329-L335):
```ts
CallableFunction.apply<
  T,
  A extends any[],
  R,
>(
  this: (
    this: T,
    ...args: A,
  ) => R,
  thisArg: T,
  args: A,
): R;
```

Using intellisense, we can see that it resolves the following types in the generic call:

```ts
CallableFunction.apply<
  ProvidedFunction,
  Parameters<ProvidedFunction>,
  any,
>(
  this: (
    this: ProvidedFunction,
    ...args: Parameters<ProvidedFunction>,
  ) => any,
  thisArg: ProvidedFunction,
  args: Parameters<ProvidedFunction>,
): any
```

Note that the 3rd generic argument has been resolved to `any` and not `ReturnType<ProvidedFunction>`, unlike the second parameter which resolves to `Parameters<ProvidedFunction>`.
So why is TS resolving the 2nd argument to the fancy generic type, but the 3rd argument to just `any`!?!
There's no magic going on here. On line 6 you explicitly typed `arguments_` as `Parameters<ProvidedFunction>` - so TS is just using the types you told it to based on the arguments you're passing.

To illustrate the point - if we remove the explicitly type `arguments_`, and instead just use the untyped `arguments` global [then we find the signature that TS instead falls to is this one](https://github.com/microsoft/TypeScript/blob/c17fd8c479c58795253b25e7272ae9072d68a9a5/lib/lib.es5.d.ts#L273-L278):
```ts
Function.apply(
  this: Function,
  thisArg: any,
  argArray?: any,
): any;
```

So what can we do here?
Either you can disable the error with a comment and accept that working with generics is messy... or you can explicitly type your `.apply` call so the return type is correct:
```ts
const result = function_.apply<
  ProvidedFunction,
  Parameters<ProvidedFunction>,
  ReturnType<ProvidedFunction>,
>(this, arguments_);
```

This will type `result` as `ReturnType<ProvidedFunction>`, which will then not trigger the rule.
